lahja/tools/drivers/driver.py

```diff
@@ -0,0 +1,328 @@
+        return self._fn(engine, **self._kwargs)
+
+
+class SyncAction:
+    def __str__(self) -> str:
+        return self._fn.__name__
+
+    def __init__(self, fn: Callable[..., None], **kwargs: Any) -> None:
+        self._fn = fn
+        self._kwargs = kwargs
+
+    def __call__(self, engine: EngineAPI, endpoint: EndpointAPI) -> None:
+        self._fn(engine, endpoint, **self._kwargs)
+
+
+class AsyncAction:
+    def __str__(self) -> str:
+        return self._coro.__name__
+
+    def __init__(self, coro: Callable[..., Awaitable[None]], **kwargs: Any) -> None:
+        self._coro = coro
+        self._kwargs = kwargs
+
+    async def __call__(self, engine: EngineAPI, endpoint: EndpointAPI) -> None:
+        await self._coro(engine, endpoint, **self._kwargs)
+
+
+Action = Union[SyncAction, AsyncAction]
+Driver = Callable[[EngineAPI], Awaitable[None]]
+EventHandlerFn = Callable[[EndpointAPI, BaseEvent], Any]
+
+
+def driver(initializer: Initializer,
+           *actions: Action,
+           ) -> Driver:
+    return functools.partial(drive, initializer=initializer, actions=actions)
+
+
+async def drive(engine: EngineAPI,
+                initializer: Initializer,
+                actions: Sequence[Action],
+                action_timeout: int = 1,
+                ) -> None:
+    logger.debug('STARTING DRIVER')
+    async with initializer(engine) as endpoint:
+        for idx, action in enumerate(actions):
+            logger.debug('RUNNING ACTION[%d]: %s', idx, action)
+            if isinstance(action, AsyncAction):
+                try:
+                    await engine.run_with_timeout(action, engine, endpoint, timeout=action_timeout)
+                except TimeoutError as err:
+                    raise TimeoutError(
+                        f"[{endpoint}] Timeout running action #{idx}:{action}"
+                    ) from err
+            elif isinstance(action, SyncAction):
+                action(engine, endpoint)
+            else:
+                raise TypeError(f"Unsupported action: {action}")
+
+
+#
+# EndpointAPI.serve
+#
+def _serve_endpoint(engine: EngineAPI,
+                    config: ConnectionConfig) -> AsyncContextManager[EndpointAPI]:
+    logger.debug(
+        '[%s(%s)].serve(%s)',
+        engine.endpoint_class.__name__,
+        config.name,
+        config.path,
+    )
+    return engine.endpoint_class.serve(config)
+
+
+def serve_endpoint(config: ConnectionConfig) -> Initializer:
+    return Initializer(_serve_endpoint, config=config)
+
+
+#
+# EndpointAPI.run
+#
+def _run_endpoint(engine: EngineAPI, name: str) -> AsyncContextManager[EndpointAPI]:
+    logger.debug('[%s(%s)].run()', engine.endpoint_class.__name__, name)
+    return engine.endpoint_class(name).run()
+
+
+def run_endpoint(name) -> Initializer:
+    return Initializer(_run_endpoint, name=name)
+
+
+#
+# Serve request
+#
+GetResponseFn = Callable[[EndpointAPI, BaseRequestResponseEvent], BaseEvent]
+
+
+async def _serve_request(engine: EngineAPI,
```

Comment:
Looks like `engine` can be dropped from this and some of the following functions.

Reply:
It needs to be present because all of the driver `Action` APIs have to have the same signature.  It's not used in most of them, but it's present to give access to the engine for a case where a `sleep` needs to occur.